In [1]:
import tensorflow as tf

In [2]:
from tensorflow.examples.tutorials.mnist import input_data

In [3]:
mnist = input_data.read_data_sets("./mnist/data/", one_hot=True)

Instructions for updating:
Please use alternatives such as official/mnist/dataset.py from tensorflow/models.
Instructions for updating:
Please write your own downloading logic.
Instructions for updating:
Please use urllib or similar directly.
Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting ./mnist/data/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting ./mnist/data/train-labels-idx1-ubyte.gz
Instructions for updating:
Please use tf.one_hot on tensors.
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting ./mnist/data/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting ./mnist/data/t10k-labels-idx1-ubyte.gz
Instructions for updating:
Please use alternatives such as official/mnist/datas

In [5]:
mnist

Datasets(train=<tensorflow.contrib.learn.python.learn.datasets.mnist.DataSet object at 0xb2be94828>, validation=<tensorflow.contrib.learn.python.learn.datasets.mnist.DataSet object at 0xb2be944a8>, test=<tensorflow.contrib.learn.python.learn.datasets.mnist.DataSet object at 0x1a2f0a3cc0>)

In [9]:
type(mnist.train)

tensorflow.contrib.learn.python.learn.datasets.mnist.DataSet

In [8]:
type(mnist.train.images)

numpy.ndarray

In [13]:
mnist.train.images.shape

(55000, 784)

In [15]:
mn = mnist.train.images.reshape(55000, 28,28,1)

In [17]:
mn[0].shape

(28, 28, 1)

In [19]:
mn[0][0].shape

(28, 1)

In [18]:
mn[0][0]

array([[0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.]], dtype=float32)

In [21]:
mn[0,0,0].shape

(1,)

In [22]:
mn[0,0,0]

array([0.], dtype=float32)

In [23]:
mn[0,0,0,0]

0.0

## 신경망 모델 구성

### None은 데이터 개수,  28*28*1 은 784를 3차원으로 만듬

In [29]:
s = mnist.train.images[0]

In [30]:
s.shape

(784,)

In [31]:
s = s.reshape(28,28,1)

In [32]:
s.shape

(28, 28, 1)

### 입력 데이터

In [40]:
X = tf.placeholder(tf.float32, [None, 28,28,1])

## 결과데이터

In [10]:
Y = tf.placeholder(tf.float32, [None, 10])

### 드롭아웃 데이터

In [11]:
keep_prob = tf.placeholder(tf.float32)

# 2. 레이어 구성

## 컨볼루션 레이어 구성

In [38]:
W1 = tf.Variable(tf.random_normal([3,3,1,32], stddev=0.01))

Instructions for updating:
Colocations handled automatically by placer.


In [41]:
L1 = tf.nn.conv2d(X, W1, strides=[1,1,1,1], padding='SAME')

In [42]:
L1 = tf.nn.relu(L1)

## 풀링 레이어 구성

In [43]:
L1 = tf.nn.max_pool(L1, ksize=[1,2,2,1], strides=[1,2,2,1], padding='SAME')

In [46]:
W2 = tf.Variable(tf.random_normal([3,3,32,64], stddev=0.01))

In [47]:
L2 = tf.nn.conv2d(L1, W2, strides=[1,1,1,1], padding='SAME')

In [48]:
L2 = tf.nn.relu(L2)

In [49]:
L2 = tf.nn.max_pool(L2, ksize=[1,2,2,1], strides=[1,2,2,1], padding='SAME')

In [50]:
W3 = tf.Variable(tf.random_normal([7*7*64,256], stddev=0.01))

In [51]:
L3 = tf.reshape(L2, [-1, 7*7*64])

In [53]:
L3 = tf.matmul(L3,  W3) 

In [54]:
L3 = tf.nn.relu(L3)

In [55]:
L3 = tf.nn.dropout(L3, keep_prob)

In [56]:
W4 = tf.Variable(tf.random_normal([256,10], stddev=0.01))

In [58]:
model = tf.matmul(L3, W4)

In [59]:
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=model, labels=Y))

In [60]:
optimizer = tf.train.AdamOptimizer(0.01).minimize(cost)

# 3. 학습

In [61]:
init = tf.global_variables_initializer()

In [62]:
sess = tf.Session()

In [63]:
sess.run(init)

In [64]:
batch_size = 100

In [65]:
total_batch = int(mnist.train.num_examples / batch_size)

In [66]:
for epoch in range(15) :
    total_cost = 0
    
    for i in range(total_batch) :
        batch_xs, batch_ys = mnist.train.next_batch(batch_size)
        batch_xs = batch_xs.reshape(-1,28,28,1)
        _, cost_val = sess.run([optimizer, cost], feed_dict={X : batch_xs,
                                                                                        Y : batch_ys,
                                                                                        keep_prob: 0.7})
        
        total_cost += cost_val
    
    print('Epoch', "%04d" % (epoch + 1), ' Avg cost =', "{:.3f}".format(total_cost/total_batch))
    
print(' 최적화 완료 ')

Epoch 0001  Avg cost = 0.243
Epoch 0002  Avg cost = 0.100
Epoch 0003  Avg cost = 0.087
Epoch 0004  Avg cost = 0.083
Epoch 0005  Avg cost = 0.076
Epoch 0006  Avg cost = 0.067
Epoch 0007  Avg cost = 0.072
Epoch 0008  Avg cost = 0.069
Epoch 0009  Avg cost = 0.066
Epoch 0010  Avg cost = 0.071
Epoch 0011  Avg cost = 0.067
Epoch 0012  Avg cost = 0.072
Epoch 0013  Avg cost = 0.065
Epoch 0014  Avg cost = 0.067
Epoch 0015  Avg cost = 0.058
 최적화 완료 


# 4. 결과 확인 


In [70]:
is_correct = tf.equal(tf.argmax(model,1), tf.argmax(Y, 1))

In [71]:
accuracy = tf.reduce_mean(tf.cast(is_correct, tf.float32))

In [72]:
print(" 정확도 ", sess.run(accuracy, feed_dict = { X : mnist.test.images.reshape(-1, 28, 28,1),
                                                                            Y : mnist.test.labels,
                                                                            keep_prob: 1}))

 정확도  0.9892
